### PR TITLE
Tags feature

### PR DESCRIPTION
**Pull request by tomeroth**

### CODE REVIEW

File: app/controllers/api/work_times_controller.rb [hunk 1]
```diff
@@ -115,13 +118,34 @@ def find_work_time
...
+                tag
+                starts_at
+                ends_at
+              ]
+            )
     end
 
     def work_time_params
-      params.fetch(:work_time).permit(:project_id, :body, :task, :starts_at, :ends_at)
+      params.fetch(:work_time)
```
Comment (mhib) [inline app/controllers/api/work_times_controller.rb]: Why not `require`?

File: app/javascript/components/shared/work_time_tag.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const WorkTimeTag = ({
+  onClick, children, workTime, editing,
+}) => {
+  if (editing === true) {
```
Comment (mhib) [inline app/javascript/components/shared/work_time_tag.js]: :stuck_out_tongue:

File: app/javascript/components/shared/work_time_tag.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    return children;
+  } if (workTime.tag === 'dev') {
```
Comment (mhib) [inline app/javascript/components/shared/work_time_tag.js]: Consider aligning return next to condition in one-line ifs

File: app/javascript/components/time_table/timesheet/entry.js [hunk 4]
```diff
@@ -156,12 +163,21 @@ class Entry extends React.Component {
           if (e.errors.task) newErrors.task = e.errors.task;
           this.setState({ errors: newErrors });
         } else {
-          alert(I18n.t('activerecord.errors.models.work_time.basic'));
+          // alert(I18n.t('activerecord.errors.models.work_time.basic'));
```
Comment (mhib) [inline app/javascript/components/time_table/timesheet/entry.js]: Why?

File: app/javascript/components/time_table/timesheet/history/entry_history.js [hunk 5]
```diff
@@ -334,6 +348,27 @@ class EntryHistory extends React.Component {
     }
   }
 
+  onTagFilter(e) {
+    e.preventDefault();
+    const { from, to } = this.state;
+    const sTag = e.target.attributes.getNamedItem('data-tag').value;
+    if (sTag) {
+      this.setState({
+        selectedTag: _.find(this.props.tags, tag => tag.key === sTag).value,
```
Comment (mhib) [inline app/javascript/components/time_table/timesheet/history/entry_history.js]: I think that Array's find would work as well

File: app/javascript/components/time_table/timesheet/history/entry_history.js [hunk 6]
```diff
@@ -461,6 +498,16 @@ class EntryHistory extends React.Component {
             )) }
           </div>
         </div>
+        <div className="button dropdown right floated scrolling teal ui" tabIndex="0">
+          <div className="text">{selectedTag ? selectedTag : I18n.t('apps.timesheet.select_tag')}</div>
```
Comment (mhib) [inline app/javascript/components/time_table/timesheet/history/entry_history.js]: `selectedTag || I18n.t('apps.timesheet.select_tag')` is less verbose

File: app/javascript/components/time_table/timesheet/history/work_hours.js [hunk 7]
```diff
@@ -128,6 +132,10 @@ class WorkHours extends React.Component {
     });
   }
 
+  onTagChange(value) {
+    this.setState({ tag: value });
```
Comment (mhib) [inline app/javascript/components/time_table/timesheet/history/work_hours.js]: consider renaming `value` to `tag` so so you can replace it with `this.setState({ tag })`

File: app/javascript/components/time_table/timesheet/history/work_hours.js [hunk 8]
```diff
@@ -317,6 +341,9 @@ class WorkHours extends React.Component {
                   ) : null }
               </div>
             </div>
+            <WorkTimeTag editing={editing} workTime={workHours} onClick={this.toggleEdit}>
+              { editing ? this.renderTagEditable() : null }
```
Comment (mhib) [inline app/javascript/components/time_table/timesheet/history/work_hours.js]: `editing && this.renderTagEditable()` is less verbose

File: app/javascript/components/time_table/timesheet/tags_list.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+          if (tag.key === 'dev') {
+            return null;
+          }
+          return (
+            <button
+              className={tag.key === selectedTag ? 'btn btn-success' : 'btn btn-info'}
+              onClick={this.onChangeTag.bind(this, tag.key)}
+              name="tag-item"
+              type="button"
+              is-selected={tag.key === selectedTag ? 'true' : 'false'}
```
Comment (mhib) [inline app/javascript/components/time_table/timesheet/tags_list.js]: Why is string needed here?

File: app/javascript/components/time_table/timesheet/tags_list.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+class TagsList extends React.Component {
```
Comment (mhib) [inline app/javascript/components/time_table/timesheet/tags_list.js]: Consider refactoring to function as this component is stateless.

**Comment by mhib** [comment]
Review done

File: app/javascript/components/time_table/timesheet/tags_list.js (hunk 9 above)
Comment (tomeroth) [reply app/javascript/components/time_table/timesheet/tags_list.js]: changed to '1 : 0', DOM would not accept boolean values

File: app/javascript/components/time_table/timesheet/history/entry_history.js (hunk 5 above)
Comment (tomeroth) [reply app/javascript/components/time_table/timesheet/history/entry_history.js]: We don't need filtering here in entry_history so this line won't exist anyways

File: app/controllers/api/work_times_controller.rb (hunk 1 above)
Comment (tomeroth) [reply app/controllers/api/work_times_controller.rb]: I just extended the permit part, didn't even noticed fetch. Ofc should be required

File: app/javascript/components/time_table/timesheet/history/work_hours.js [hunk 11]
```diff
@@ -208,20 +237,37 @@ class WorkHours extends React.Component {
     });
   }
 
+  toggleTagEdit() {
```
Comment (mhib) [inline app/javascript/components/time_table/timesheet/history/work_hours.js]: As state is edited based on itself here I'd use setState with updater function instead.

File: app/assets/stylesheets/timesheet/blocks/time_entry.sass [hunk 12]
```diff
@@ -173,7 +173,16 @@
     -webkit-flex: 0 1 auto
     -ms-flex: 0 1 auto
     flex: 0 1 auto
-    padding: 16px 0
+    padding: 16px 10px
+
+  .tag-container
+    overflow: hidden
+    max-width: 200px
+    -webkit-box-flex: 1
```
[the remark this answers is not in the document]
Comment (tomeroth) [reply app/assets/stylesheets/timesheet/blocks/time_entry.sass]: Dunno, just made it the same way other classes were made. Have not found any

[the remark this answers is not in the document]
Comment (mhib) [reply app/assets/stylesheets/timesheet/blocks/time_entry.sass]: :man_shrugging:

**Pull request merged**